macosx/FloatEqual.h

```diff
@@ -0,0 +1,15 @@
+// This file Copyright © 2022 Transmission authors and contributors.
+// It may be used under the MIT (SPDX: MIT) license.
+// License text can be found in the licenses/ folder.
+
+#import <CoreFoundation/CoreFoundation.h>
+
+static inline BOOL FloatEqual(CGFloat a, CGFloat b)
+{
+    // Comparing values of same type is safe;
+    // The `float-equal` warning should not be used.
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wfloat-equal"
+    return a == b;
```

Comment:
<details><summary>Proof to the contrary</summary>

Code:
```cpp
#include <iostream>
#include <iterator>
#include <utility>

#define ITEM(x) {#x, x}

int main() {
    std::pair<char const*, double> items[] = {
        ITEM(0.59),
        ITEM(0.59F),
        ITEM(1.59 - 1),
        ITEM(1000.59 - 1000),
    };

    std::cout << "items:" << std::endl;
    for (size_t i = 0; i < std::size(items); ++i) {
        std::cout << "  " << items[i].first << " = " << items[i].second << std::endl;
    }

    std::cout << "equality:" << std::endl;
    for (size_t i = 0; i < std::size(items); ++i) {
        for (size_t j = i + 1; j < std::size(items); ++j) {
            if (items[i].second != items[j].second) {
                std::cout << "  " << items[i].first << " != " << items[j].first << std::endl;
            }
        }
    }
}
```

Output:
```
items:
  0.59 = 0.59
  0.59F = 0.59
  1.59 - 1 = 0.59
  1000.59 - 1000 = 0.59
equality:
  0.59 != 0.59F
  0.59 != 1.59 - 1
  0.59 != 1000.59 - 1000
  0.59F != 1.59 - 1
  0.59F != 1000.59 - 1000
  1.59 - 1 != 1000.59 - 1000
```
</details>

<details><summary>Examples of existing checks</summary>

https://github.com/transmission/transmission/blob/e694c3e3a00ae5dd65e7273fcc7f3d3f369a51c1/gtk/Session.cc#L1323-L1339

https://github.com/transmission/transmission/blob/e694c3e3a00ae5dd65e7273fcc7f3d3f369a51c1/qt/VariantHelpers.cc#L22-L32

https://github.com/transmission/transmission/blob/e694c3e3a00ae5dd65e7273fcc7f3d3f369a51c1/tests/libtransmission/settings-test.cc#L63-L76
</details>

Reply:
Those are examples with different types (0.59, 0.59F) or with subtractions. The usage in our code is to detect if a value has been modified or not. `-Wfloat-equal` doesn't allow observing if a floating value has changed. I'll update the comment.

Reply:
The type is exactly the same when comparison is taking place. The function doesn't (and can't) prevent people from "misusing" it but gives the false impression that it'll work in all the example cases.

Reply:
The function is documented.
The present code usage is correct.
And well, it's this or adding `-Wno-float-equal`, not much alternatives.

Reply:
I think the discussion of double vs float is not the central point here; even comparison of two floats with the equality operator is still unsafe. Surely we need to compare the difference to `std::numeric_limits<float>::epsilon();` similar to the VariantHelpers code?

Reply:
There are no maths operations:
- we assign 123.45678 to _storage and to _backupStorage.
- we can't observe changes to _storage directly.
- regularly, we check if _storage != _backupStorage or not.

Reply:
As the Mandalorian would say, "_this is the way_":
https://stackoverflow.com/questions/65939711/setting-and-checking-with-a-float-constant

Reply:
@mikedld kindly inform me of your desired change, if any, so that we can move forward and get the various changes merged. Or if you want some alternative workaround like `a >= b && a <= b` or `std::equal_to<FloatType>()(a, b)`, then I'll abide by it.

Reply:
> There are no arithmetics operations

Well, for example we have

```
    CGFloat difference = self.fHeightChange;
    if (!FloatEqual(difference, CGFloat(0)))
````

Where `fHeightChange` is:

```
- (CGFloat)fHeightChange
{
    return self.oldHeight - self.fCurrentHeight;
}
```

This seems like FloatEqual() is being called with inputs that come from arithmetic operations?

Reply:
No, it's the old value and the new value: same as passing `FloatEqual(self.oldHeight, self.fCurrentHeight)`

Reply:
I think I'm not understanding what you're saying?

You're saying there are no arithmetic operations, but surely `self.oldHeight - self.fCurrentHeight` is an arithmetic operation whose result used as the first argument to a call to FloatEqual()?

Reply:
`FloatEqual(a, b)` is same as `FloatEqual(a - b, 0)`.

Reply:
OK. I'm still confused but our area of disagreement is smaller, so this is progress :smile_cat: 

My understanding is that  `a - b` is an arithmetic operation and so its result would not be safe for use with `==`. Is that incorrect? And if so, why?

Reply:
How do you code that:
>I want to know when A and B are perfectly equal, and I want to know their difference when they aren't.

The type of arithmetics that is unsafe is "_Is 1.6 - 0.6 equal to 1.0 ?_" because of rounding. But doing "_A - A_" is always going to be 0.

Reply:
I don't understand how `a - b` isn't the same as the `1.6 - 0.6` case here.

The fact that the right-hand size is zero doesn't seem relevant since the math operations on the left-hand side may have already introduced epsilon drift _before_ being compared to the right-hand side?

Reply:
This is taking time to explain.
I think I should have used the word **computations** instead of **arithmetics** to explain it.
Demonstrations that there are valid cases for PERFECT float equality are:
- https://stackoverflow.com/questions/51134021/floating-point-equality
- https://stackoverflow.com/questions/65939711/setting-and-checking-with-a-float-constant

In the mean time, I'm splitting the PR: all the changes except FloatEqual are in #4414.

Reply:
It's only modified if the size of the UI has changed and needs to be redrawn. Precision is likely at most 0.1 pixels on some Retina displays.

Reply:
Just leave it here: https://source.chromium.org/chromium/chromium/src/+/main:ui/base/cocoa/appkit_utils.mm;l=38-40

nit: As we have a lot comparisions with zero, `FloatIsZero` may be viable also.

Reply:
If 2 * a != 2 * b, then fabs(a * 2 - b * 2) is going to be greater than epsilon. So Chromium way has its purpose, while not being a universal solution.
In our case, we don't need to compare with epsilon: we only care about perfect equality.

Reply:
http://realtimecollisiondetection.net/blog/?p=89 discusses that weakness in the Chromium approach and has a nice (and well-described) revision to try to address its weaknesses. What would you think about using the approach in that blog post here?

Reply:
That approach would be only for Chromium case, but that would degrade their performances. I would best not use that function at all.